File: exercises/day10-TimePortal/typescript/tests/ObjectCalisthenicsRules.test.ts
import * as fs from 'fs';
import * as path from 'path';

/**
 * 🌀 CHRONOS - Time Portal Activation Sequence
 * 
 * These tests check Object Calisthenics compliance by analyzing the source code.
 * Since there's no ArchUnit for TypeScript, we use simple source analysis.
 */

const srcDir = path.join(__dirname, '../src');

function readSourceFile(filename: string): string {
    return fs.readFileSync(path.join(srcDir, filename), 'utf-8');
}

describe('🌀 CHRONOS - Time Portal Activation Sequence', () => {

    beforeAll(() => {
        console.log(`
╔═══════════════════════════════════════════════════════════════╗
║  🌀 CHRONOS TEMPORAL GUARDIAN SYSTEM v3.14159                 ║
║  ─────────────────────────────────────────────────────────── ║
║  Scanning codebase for Object Calisthenics compliance...     ║
╚═══════════════════════════════════════════════════════════════╝
        `);
    });

    describe('⚡ Energy Containment Protocol', () => {

        test('🔋 Energy levels must be encapsulated in a value object', () => {
            const source = readSourceFile('TimePortal.ts');
            const hasPrimitiveEnergy = /_energyLevel:\s*number/.test(source);
            
            expect(hasPrimitiveEnergy).toBe(false);
            // Fails because _energyLevel is a primitive number
            // Create an Energy value object/class instead
        });

        test('📅 Year fields must be wrapped in a Year value object', () => {
            const travelerSource = readSourceFile('Traveler.ts');
            const hasPrimitiveYear = /_currentYear:\s*number/.test(travelerSource);
            
            expect(hasPrimitiveYear).toBe(false);
            // Fails because _currentYear is a primitive number
            // Create a Year value object/class instead
        });
    });

    describe('📜 Temporal Record Keeping', () => {

        test('📚 Travel logs must be encapsulated in a dedicated TravelLog type', () => {
            const source = readSourceFile('TimePortal.ts');
            const hasRawArray = /_travelLog:\s*string\[\]/.test(source);
            
            expect(hasRawArray).toBe(false);
            // Fails because _travelLog is a raw string array
            // Create a TravelLog first-class collection instead
        });

        test('🌐 Portal collections must be wrapped in a PortalRegistry', () => {
            const source = readSourceFile('PortalNetwork.ts');
            const hasRawArray = /_portals:\s*TimePortal\[\]/.test(source);
            
            expect(hasRawArray).toBe(false);
            // Fails because _portals is a raw array
            // Create a PortalRegistry first-class collection instead
        });
    });

    describe('🔒 Collection Integrity Shield', () => {

        test('🛡️ No getter should expose raw array types', () => {
            const portalSource = readSourceFile('TimePortal.ts');
            const networkSource = readSourceFile('PortalNetwork.ts');
            
            const exposesRawTravelLog = /get\s+travelLog\(\):\s*string\[\]/.test(portalSource);
            const exposesRawPortals = /get\s+portals\(\):\s*TimePortal\[\]/.test(networkSource);
            
            expect(exposesRawTravelLog || exposesRawPortals).toBe(false);
            // Fails because raw arrays are exposed
            // Return immutable views or domain-specific types instead
        });
    });

    describe('🎯 Status Type Safety', () => {

        test('⚡ Portal status must use an enum, not raw string', () => {
            const source = readSourceFile('TimePortal.ts');
            const hasStringStatus = /_currentStatus:\s*string/.test(source);
            
            expect(hasStringStatus).toBe(false);
            // Fails because _currentStatus is a raw string
            // Create a PortalStatus enum instead
        });

        test('❤️ Health status must use an enum, not raw string', () => {
            const source = readSourceFile('Traveler.ts');
            const hasStringHealth = /_healthStatus:\s*string/.test(source);
            
            expect(hasStringHealth).toBe(false);
            // Fails because _healthStatus is a raw string
            // Create a HealthStatus enum instead
        });
    });

    describe('🚀 FINAL ACTIVATION SEQUENCE', () => {

        test('🌀 All systems nominal - Portal ready for temporal operations', () => {
            const source = readSourceFile('TimePortal.ts');
            
            // Check that TimePortal has no primitive number fields
            const hasPrimitiveNumber = /_energyLevel:\s*number/.test(source);
            
            expect(hasPrimitiveNumber).toBe(false);

            console.log(`
╔═══════════════════════════════════════════════════════════════╗
║  🌀 CHRONOS: ALL SYSTEMS NOMINAL                             ║
║  ─────────────────────────────────────────────────────────── ║
║  ✅ Energy containment: STABLE                               ║
║  ✅ Temporal records: SECURED                                ║
║  ✅ Collection integrity: PROTECTED                          ║
║  ✅ Status encoding: TYPE-SAFE                               ║
║                                                               ║
║  🚀 PORTAL ACTIVATION: AUTHORIZED                            ║
╚═══════════════════════════════════════════════════════════════╝
            `);
        });
    });
});

File: exercises/day10-TimePortal/scala/src/test/scala/timeportal/ObjectCalisthenicsRulesTest.scala
package timeportal

import com.tngtech.archunit.core.importer.ClassFileImporter
import com.tngtech.archunit.lang.syntax.ArchRuleDefinition.{noFields, noMethods}
import org.scalatest.BeforeAndAfterAll
import org.scalatest.funspec.AnyFunSpec
import org.scalatest.matchers.should.Matchers

class ObjectCalisthenicsRulesTest extends AnyFunSpec with Matchers with BeforeAndAfterAll:

  private val classes = new ClassFileImporter().importPackages("timeportal")

  override def beforeAll(): Unit =
    println(
      """
        |╔═══════════════════════════════════════════════════════════════╗
        |║  🌀 CHRONOS TEMPORAL GUARDIAN SYSTEM v3.14159                 ║
        |║  ─────────────────────────────────────────────────────────── ║
        |║  Scanning codebase for Object Calisthenics compliance...     ║
        |╚═══════════════════════════════════════════════════════════════╝
        |""".stripMargin)

  describe("⚡ Energy Containment Protocol"):

    it("🔋 Energy levels must be encapsulated in a value object"):
      val rule = noFields()
        .that().haveName("_energyLevel")
        .should().haveRawType(classOf[Int])
        .because("Portal energy requires proper encapsulation to prevent temporal leaks. Create an Energy value object.")

      rule.check(classes)

    it("📅 Year fields must be wrapped in a Year value object"):
      val rule = noFields()
        .that().haveNameMatching(".*[Yy]ear.*")
        .should().haveRawType(classOf[Int])
        .because("Temporal coordinates require Year value objects to enforce boundaries (-10000 to 3000). Create a Year value object.")

      rule.check(classes)

  describe("📜 Temporal Record Keeping"):

    it("📚 Travel logs must be encapsulated in a dedicated TravelLog type"):
      val rule = noFields()
        .that().haveNameMatching(".*[Ll]og.*")
        .should().haveRawType(classOf[scala.collection.mutable.ListBuffer[?]])
        .because("Travel history requires a TravelLog collection class for proper temporal record keeping. Raw ListBuffer exposes internal structure.")

      rule.check(classes)

    it("🌐 Portal collections must be wrapped in a PortalRegistry"):
      val rule = noFields()
        .that().haveName("portals")
        .should().haveRawType(classOf[scala.collection.mutable.ListBuffer[?]])
        .because("Portal networks require a dedicated PortalRegistry for proper management. Raw ListBuffer<TimePortal> lacks domain behavior.")

      rule.check(classes)

  describe("🔒 Collection Integrity Shield"):

    it("🛡️ No method should expose raw List types"):
      val rule = noMethods()
        .that().haveNameMatching("get.*")
        .and().areDeclaredInClassesThat().resideInAPackage("timeportal")
        .should().haveRawReturnType(classOf[List[?]])
        .because("Exposing raw collections allows external temporal interference. Return immutable views or domain-specific types instead.")

      rule.check(classes)

  describe("🎯 Status Type Safety"):

    it("⚡ Portal status must use an enum, not raw String"):
      val rule = noFields()
        .that().haveName("_currentStatus")
        .should().haveRawType(classOf[String])
        .because("Status indicators require type-safe enums. Create a PortalStatus enum with IDLE, CHARGING, JUMPING, ERROR values.")

      rule.check(classes)

    it("❤️ Health status must use an enum, not raw String"):
      val rule = noFields()
        .that().haveName("_healthStatus")
        .should().haveRawType(classOf[String])
        .because("Traveler health requires type-safe status. Create a HealthStatus enum with STABLE, UNSTABLE, CRITICAL values.")

      rule.check(classes)

  describe("🚀 FINAL ACTIVATION SEQUENCE"):

    it("🌀 All systems nominal - Portal ready for temporal operations"):
      val rule = noFields()
        .that().areDeclaredInClassesThat().haveSimpleName("TimePortal")
        .should().haveRawType(classOf[Int])
        .because("TimePortal must have no primitive int fields for stable temporal operations.")

      rule.check(classes)

      println(
        """
          |╔═══════════════════════════════════════════════════════════════╗
          |║  🌀 CHRONOS: ALL SYSTEMS NOMINAL                             ║
          |║  ─────────────────────────────────────────────────────────── ║
          |║  ✅ Energy containment: STABLE                               ║
          |║  ✅ Temporal records: SECURED                                ║
          |║  ✅ Collection integrity: PROTECTED                          ║
          |║  ✅ Status encoding: TYPE-SAFE                               ║
          |║                                                               ║
          |║  🚀 PORTAL ACTIVATION: AUTHORIZED                            ║
          |╚═══════════════════════════════════════════════════════════════╝
          |""".stripMargin)
